src/main/java/com/google/research/bleth/services/DatabaseService.java

```diff
@@ -0,0 +1,157 @@
+package com.google.research.bleth.services;
+
+import com.google.appengine.api.datastore.*;
+import com.google.research.bleth.simulator.Board;
+
+/**
+ * DatabaseService is a singleton class wrapping the application's datastore instance.
+ * DatabaseService provides a global point of access to any read/write operation using the datastore instance.
+ */
+public class DatabaseService {
+
+    private static DatabaseService instance = null;
+    private DatastoreService datastore;
+
+    /**
+     * A private constructor.
+     * Ensures that the only way to create an instance of the class is by calling the getInstance() static method.
+     */
+    private DatabaseService() {
+        datastore = DatastoreServiceFactory.getDatastoreService();
+    }
+
+    /**
+     * Static method to create an instance of the DatabaseService class.
+     * @return a DatabaseService class instance.
+     */
+    public static DatabaseService getInstance() {
+        if (instance == null) {
+            instance = new DatabaseService();
+        }
+        return instance;
+    }
+
+    private void writeBoardState(String simulationId, int round, Board board, String entityKind)
+            throws PreparedQuery.TooManyResultsException {
+
+        // Validate there is no such board state in datastore.
+        if (this.getBoardState(simulationId, round, entityKind) != null) {
+            throw new PreparedQuery.TooManyResultsException();
+        }
+
+        // Create new entity.
+        Entity boardStateEntity = new Entity(entityKind);
+
+        // Set properties.
+        boardStateEntity.setProperty("simulationId", simulationId);
+        boardStateEntity.setProperty("round", round);
+        boardStateEntity.setProperty("state", BoardParser.parse(board));
+
+        // Write to datastore.
+        datastore.put(boardStateEntity);
+    }
+
+    /**
+     * Create an entity of a real board state based on input params and write it to datastore.
+     * @param simulationId is the Id of the simulation associated with the input board state.
+     * @param round is the round number in the simulation that matches the input board state.
+     * @param board is a Board object representing the state of the board.
+     * @throws PreparedQuery.TooManyResultsException if a board state with the same simulationId and round already exists.
+     */
+    public void writeRealBoardState(String simulationId, int round, Board board)
```

Comment:
This sounds like a great solution for the write method, I'm on it.

But as for reading I believe we will still have two separate methods, because we will most likely have both real and estimated entities for the same combination of simulation ID and round number stored on the db.

An alternative could be having a single method that gets a flag for real or estimated as an argument, but I recall you've already mentioned this solution as a bad practice.

Reply:
Is an attribute of typeBoard (enum which can be REAL or ESTIMATED) good enough for this purpose? 
Is there any benefit in making Board abstract as you suggested?